Added modifications for lensingGW's usage and new lens profiles

gipagano opened the pull request.

sibirrer commented on lenstronomy/LensModel/Profiles/diego_et_al_profile.py:
```diff
@@ -0,0 +1,103 @@
+import numpy as np
+from lenstronomy.LensModel.Profiles.base_profile import LensProfileBase
+
+class DiegoGalaxy(LensProfileBase):
```
This profile is effectively a constant Hessian matrix with a parametrization in terms of radial and tangential stretch parameters. I am not sure this is recognizable for the user as 'DiegoGalaxy'. What about 'ConstMag'?

To have the full Hessian matrix being explored, a direction (e.g. of mu_r) relative to x vs y is required.

sibirrer commented on lenstronomy/LensModel/Profiles/diego_et_al_profile.py:
```diff
@@ -0,0 +1,103 @@
+    """
+    
+    param_names = ['center_x', 'center_y','mu_r', 'mu_t', 'parity']
+    lower_limit_default = {'center_x': -100, 'center_y': -100}
```
the default lower and upper limits should be defined for all parameters. This is not really relevant for you but a user that wants to fit a parameterized model to some data might request these default values in the initialization of the fitting sequence.

gipagano replied:
Done

sibirrer commented on lenstronomy/LensModel/Profiles/diego_et_al_profile.py:
```diff
@@ -0,0 +1,103 @@
+            kappa  = 1 -gamma1+1./mu_r
+        else:
+            raise ValueError('%f is not a valid value for the parity of the macromodel. Chose either +1 or -1.' % parity)
+        x_shift = x - center_x
```
for building in rotation, you can e.g. look at the NIE profile in lenstronomy. Effectively you can rotate the coordinates, then evaluate the quantities and then rotate back based on the symmetry of the quantity (the potential is a scalar and does not need to be rotated)

sibirrer replied:
alternatively to calculating gamma1, gamma2, kappa, you can transform mu_r, mu_t into f_xx, f_yy, f_xy of the Hessian matrix. Just a thought.

sibirrer commented on lenstronomy/LensModel/Profiles/elliptical_profile.py:
```diff
@@ -0,0 +1,66 @@
+import numpy as np
+from lenstronomy.LensModel.Profiles.base_profile import LensProfileBase
+
+class Elliptical(LensProfileBase):
```
I always struggle with naming, but it's important to be recognizable. This is an elliptical isothermal potential. What about 'NIE_POTENTIAL'? 'NIE' stands for non-singular Isothermal Ellipse and is a well known and often used model. The 'NIE' is already implemented in lenstronomy and it would give a nice correspondence.

The same as for the Diego et al. model, orientation angle is not implemented. To perform modeling, it is convenient to define the ellipticity (incl direction) with eccentricity parameters e1, e2. I think giving the user the possibility for choosing the direction is important. Have a look at the NIE model, in case.

For completeness, you can also give the reference to
Blandford & Kochanek 1987

sibirrer commented on lenstronomy/LensModel/Solver/lens_equation_solver.py:
```diff
@@ -68,6 +68,39 @@ def _root(self, x, kwargs_lens, source_x, source_y):
         beta_x, beta_y = self.lensModel.ray_shooting(x_, y_, kwargs_lens)
         return (beta_x - source_x)**2 + (beta_y - source_y)**2
+        
+    def candidate_solutions(self, sourcePos_x, sourcePos_y, kwargs_lens, min_distance=0.1, search_window=10,
```
this is well done!

gipagano replied:
Thanks!

sibirrer commented on lenstronomy/LensModel/profile_list_base.py:
```diff
@@ -55,6 +55,12 @@ def _import_class(lens_type, custom_class, z_lens=None, z_source=None):
         if lens_type == 'SHIFT':
             from lenstronomy.LensModel.Profiles.alpha_shift import Shift
             return Shift()
+        elif lens_type == 'ELLIPTICAL':
```
naming as mentioned

gipagano replied:
Done

sibirrer commented on lenstronomy/LensModel/profile_list_base.py:
```diff
@@ -55,6 +55,12 @@ def _import_class(lens_type, custom_class, z_lens=None, z_source=None):
+            from lensinggw.examples.potentials.elliptical_profile import Elliptical
+            return Elliptical()
+        elif lens_type == 'DIEGO_GALAXY':
+            from lensinggw.examples.potentials.diego_et_al_profile import DiegoGalaxy
```
naming as mentioned

sibirrer replied:
Hi Giulia, this is a great PR, thank you very much!
The code is well written, and appropriately documented such that I can follow the execution conveniently.

I have some requests for naming and the implementation of ellipticity orientation. You may say, hmm, I don't need it, that's fine. But then the next person using it and requests it, and then the updated implementation will make things less compatible with your usage.

Another thing I request for PR's is that all the lines of code are tested. In this PR, for some reason, travis-ci (integrated testing routine) did not run (I think I fixed it for your next update) but next time you will see a test coverage report.

How does testing work? lenstronomy is effectively mirrored in the /test folder.

- One test I'd like you to perform is add your new profiles to the list of models being numerically differentiated to assess the consistency between potential, deflection and hessian. This can be done in test/test_LensModel/test_numerical_lens_differentials.

- Then you can make a new file in the mirrored test_Profiles/ folder with the name test_<file_name>.py and add routines that let you assert and confirm the meaning you assigned to the functions. These can be very simple. It's a bit a hassle at first but everyone editing your function later on will have to satisfy these tests. It's important to having a stable development process.

you can locally execute the test functions by typing py.test
Or for a specific file py.test test/test_LensModel/test_numerical_lens_differentials.py

Let me know if you have questions! If you are too tiered for doing the testing, I can take it over (but then I would have to get back to you if it shows an undesirable feature).

Thank you very much!
Simon

gipagano replied:
Hi Simon,

One check of the travis CI has failed, but from the log  I can't understand what it depends on.

sibirrer commented on test/test_LensModel/test_numeric_lens_differentials.py:
```diff
@@ -275,6 +275,16 @@ def test_cored_density_2_mst(self):
         self.assert_differentials(lens_model, kwargs)
+        
+    def test_const_mag(self):
+        kwargs = {'mu_r': 1, 'mu_t': 10, 'parity': 1}
```
excellent! Can you also add a second call for parity=-1? The more testing the better :)

gipagano replied:
Added test for the negative parity side

sibirrer commented:
Hi Giulia, the travis CI build has passed all tests: https://travis-ci.org/github/sibirrer/lenstronomy
What has 'failed' is the test coverage. When you click on the courage/coveralls link, you see that the three raise statements that you nicely put into the code, where not executed and as the test coverage is high, these three lines dropped the coverage fraction. There are specific ways to test the raise statements too (which I can do later too).

gipagano commented:
Hi Simon,

I've added the orientation angle to the const_mag profile and to the nie_potential, with related tests. I've also added the static vs dynamic setting in the nie_potential, as per the nie implementation.
I've corrected a minor bug in the default upper bound of e2 in the epl profile.

Let me know if other changes are necessary.

Thanks,
Giulia

sibirrer commented on lenstronomy/LensModel/Profiles/const_mag.py:
```diff
@@ -0,0 +1,115 @@
+        :return: deflection angle (in angles)
+        """        
+        
+        # positive parity case
```
just a minor comment for these lines of code: the transforms of parity and angle into kappa, gamma1, gamma2 are duplicated for all functions in this class. Perhaps make a separate utility function processing the input. It's a very minor suggestion about coding practice not relevant for the acceptance of this PR

gipagano replied:
Thanks for the suggestion, I will implement it for the next PR
